backend/regalloc/regalloc_split.ml

```diff
@@ -132,7 +132,50 @@ let make_spill :
     ~id:(State.get_and_incr_instruction_id state)
     ~copy ~from:new_reg ~to_:stack_reg
 
-(* Inserts spills at the end of blocks, before destruction points. *)
+(* Inserts the spills in a block, as early as possible (i.e. immediately after
+   the register is last set), to reduce live ranges. *)
+let rec insert_spills_in_block :
+    State.t ->
+    block_subst:Substitution.t ->
+    stack_subst:Substitution.t ->
+    Cfg.basic_block ->
+    Instruction.t DLL.cell option ->
+    Reg.Set.t ->
+    unit =
+ fun state ~block_subst ~stack_subst block cell live_at_destruction_point ->
+  match Reg.Set.is_empty live_at_destruction_point with
+  | true -> ()
+  | false -> (
+    match cell with
+    | None ->
+      Reg.Set.iter
+        (fun old_reg ->
+          let new_reg = Substitution.apply_reg block_subst old_reg in
+          let copy = block.terminator in
+          let spill = make_spill state ~stack_subst ~old_reg ~new_reg ~copy in
+          DLL.add_begin block.body spill)
+        live_at_destruction_point
+    | Some cell ->
+      let live_at_destruction_point =
+        Reg.Set.filter
+          (fun old_reg ->
+            let new_reg = Substitution.apply_reg block_subst old_reg in
+            let instr = DLL.value cell in
+            if occurs_array instr.res new_reg
```

Comment:
should this also check that `instr` does not destroy new_reg?

Reply:
I don't think that can happen right now:
we are before register allocation, so
`old_reg` and `new_reg` would have
their locations set to `Unknown`.

Is your concern that it could happen
if we change things in _e.g._ selection?

Reply:
May be add an assert that the locationo f `new_reg` is unkown?